### PR TITLE
Single conviction doesn't block arrests on same case

### PR DESCRIPTION
**Pull request by aemrich**
Addresses #387. 

Tested for some cases in the issue. Fixes two of the reported bugs for real examples. Other reported bugs relevant to issue may be errors. There may be extra edge cases we can add tests for.

### CODE REVIEW

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 1]
```diff
@@ -133,6 +135,30 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
         for charge in charges:
             charge.set_time_ineligible(reason, eligibility_date)
 
+    @staticmethod
+    def _set_arrests_as_eligible(charges_with_summary: ChargesWithSummary):
+        # If most recent conviction is time eligible
+        # AND all other charges in the same case are arrests
+        # Then make all other charges in the case time eligible
+        mrc = charges_with_summary.most_recent_conviction
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Might as well extract out 

`second_mrc = charges_with_summary.second_most_recent_conviction`

to keep it consistent with `mrc`.

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 2]
```diff
@@ -133,6 +135,30 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+        if not mrc is None and not mrc.expungement_result.time_eligibility is None and mrc.expungement_result.time_eligibility.status is EligibilityStatus.ELIGIBLE:
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Does

`if mrc and mrc.expungement_result.time_eligibility and mrc.expungement_result.time_eligibility.status is EligibilityStatus.ELIGIBLE:`

work?

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 3]
```diff
@@ -133,6 +135,30 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+                pass
+            else:
+                for charge in mrc.case()().charges:
+                    if charge.acquitted():
+                        charge.set_time_eligible()
+        else:
+            if charges_with_summary.second_most_recent_conviction and not mrc is None:
+                for charge in mrc.case()().charges:
+                    if not charges_with_summary.second_most_recent_conviction.disposition is None:
+                        if charge.acquitted():
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: I think you can combine this to

`if second_mrc.disposition and charge.acquitted()`

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 4]
```diff
@@ -133,6 +135,30 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+    @staticmethod
+    def _set_arrests_as_eligible(charges_with_summary: ChargesWithSummary):
+        # If most recent conviction is time eligible
+        # AND all other charges in the same case are arrests
+        # Then make all other charges in the case time eligible
+        mrc = charges_with_summary.most_recent_conviction
+        if not mrc is None and not mrc.expungement_result.time_eligibility is None and mrc.expungement_result.time_eligibility.status is EligibilityStatus.ELIGIBLE:
+            if charges_with_summary.second_most_recent_conviction:
+                pass
+            else:
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: You can just do

`if not second_mrc:`

and delete the `pass` clause.

File: src/backend/tests/expunger/test_time_analyzer.py [hunk 5]
```diff
@@ -16,6 +16,118 @@ class TestSingleChargeAcquittals(unittest.TestCase):
...
+        assert three_yr_conviction.expungement_result.time_eligibility.date_will_be_eligible == date.today() + relativedelta(years=7, days=1)
+
+        assert arrest.expungement_result.time_eligibility.status is EligibilityStatus.INELIGIBLE
+        assert arrest.expungement_result.time_eligibility.date_will_be_eligible == date.today() + relativedelta(years=7)
+
+        # assert violation.expungement_result.time_eligibility.status is EligibilityStatus.INELIGIBLE
+        # assert violation.expungement_result.time_eligibility.date_will_be_eligible == date.today() + relativedelta(years=7)
+        #
+        # assert violation_2.expungement_result.time_eligibility.status is EligibilityStatus.INELIGIBLE
+        # assert violation_2.expungement_result.time_eligibility.date_will_be_eligible == date.today() + relativedelta(years=7)
```
Comment (KentShikama) [inline src/backend/tests/expunger/test_time_analyzer.py]: Looks like these comments need to be cleaned up?

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 6]
```diff
@@ -133,6 +135,28 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+        # AND all other charges in the same case are arrests
+        # Then make all other charges in the case time eligible
+        mrc = charges_with_summary.most_recent_conviction
+        second_mrc = charges_with_summary.second_most_recent_conviction
+        if mrc and mrc.expungement_result.time_eligibility and mrc.expungement_result.time_eligibility.status is EligibilityStatus.ELIGIBLE:
+            if not second_mrc:
+                for charge in mrc.case()().charges:
+                    if charge.acquitted():
+                        charge.set_time_eligible()
+        else:
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: What happens without this else block?

Comment (aemrich) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Without the else block, arrests on a case with a second mrc don't have the correct eligibility date

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 7]
```diff
@@ -133,6 +135,28 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+            if charges_with_summary.second_most_recent_conviction and not mrc is None:
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: elif mrc and second_mrc:

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py (hunk 6 above)
Comment (KentShikama) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Isn't that what line 23/24 of this file is for?

Comment (KentShikama) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: I see. I feel like there must be a better way to codify the combination of this and line 23/24 but I think this is fine for now.

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 8]
```diff
@@ -133,6 +135,28 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+                for charge in mrc.case()().charges:
+                    if second_mrc.disposition and charge.acquitted():
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Does `second_mrc.disposition` exist because of `mypy`? If so, you can just add a comment `# type: ignore` on line 156.

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 9]
```diff
@@ -133,6 +135,28 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+                            charge.set_time_ineligible(
+                                'Multiple convictions within last ten years',
+                                charges_with_summary.second_most_recent_conviction.disposition.date
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: `second_mrc.disposition.date`

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py [hunk 10]
```diff
@@ -133,6 +135,28 @@ def _mark_as_time_ineligible(charges: List[Charge], reason: str, eligibility_dat
...
+            if not second_mrc:
+                for charge in mrc.case()().charges:
+                    if charge.acquitted():
+                        charge.set_time_eligible()
+        else:
+            if charges_with_summary.second_most_recent_conviction and not mrc is None:
+                for charge in mrc.case()().charges:
+                    if second_mrc.disposition and charge.acquitted():
+                            charge.set_time_ineligible(
+                                'Multiple convictions within last ten years',
```
Comment (KentShikama) [inline src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Please reference `137.225 7(b)` somehow here.

File: src/backend/tests/expunger/test_time_analyzer.py [hunk 11]
```diff
@@ -16,6 +16,112 @@ class TestSingleChargeAcquittals(unittest.TestCase):
...
+            level='Violation',
+            case=case,
+            disposition=['Convicted', Time.THREE_YEARS_AGO])
+
+        case.charges = [three_yr_mrc, arrest, violation]
+        record = Record([case])
+        expunger = Expunger(record)
+
+        expunger.run()
+        # import pdb; pdb.set_trace()
```
Comment (KentShikama) [inline src/backend/tests/expunger/test_time_analyzer.py]: Looks like a debug comment.

**Comment by KentShikama** [comment]
As to the build errors, feel free to just add the line comment `# type: ignore` to both

```
expungeservice/expunger/analyzers/time_analyzer.py:154: error: Item "None" of "Optional[Charge]" has no attribute "disposition"
expungeservice/expunger/analyzers/time_analyzer.py:157: error: Item "None" of "Optional[Disposition]" has no attribute "date"
```

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py (hunk 6 above)
Comment (KentShikama) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: So based on what I read of 7(b), I think the charges should be set time ineligible until the mrc is 10 years old (not until the second mrc is 10 years old as is written now). @wittejm What do you think?

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py (hunk 10 above)
Comment (wittejm) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: These reason strings are all inconsistent but they can be fixed in a separate PR addressing the issue: #619

File: src/backend/expungeservice/expunger/analyzers/time_analyzer.py (hunk 6 above)
Comment (KentShikama) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: After a bit more thinking, I think this else block is unnecessary. The eligibility for everything that is not the mrc is 10 years after the mrc date, which is taken care of in line 23/24. @wittejm Any input here?

Comment (wittejm) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: If the mrc becomes eligible after the second_mrc is 10 years old, then the arrests on the same case as the mrc fall under this rule, and have the same eligibility. The reason for the `else` here is the same as the reason for having this whole fix: the arrests on this case are not blocked by the mrc, but instead only by the second_mrc

Comment (wittejm) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Though I think this  actually needs another tweak: the eligibility of the arrests on the same case as the mrc get pushed out by 10 years from the date of the second_mrc, and also by 3 years from the date of the mrc, whichever is farther out.

So for example if the mrc is 1 year old and the second_mrc is 9 years old, the arrests on the mrc's case become eligible in 2 years, not 1. 

It's confusing as hell. But I'm pretty sure there's a test case here that would be failing given the current logic.

Comment (KentShikama) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: Ah yes, I was thinking about 7(b) too strictly again. Here my high-level overview of the TimeAnalyzer without Class B Felonies as of now.

```
if mrc and second_mrc:
    # mrc and arrests in same case is eligible the max of mrc + 3 years and second_mrc + 10 years
    # everything else is eligible mrc + 10 years
elif mrc:
    # mrc and arrests in same case is eligible mrc + 3 years
    # everything else is eligible mrc + 10 years
elif mrd and second_mrd:
    # mrd is eligible second_mrd + 3 years
    # everything else is eligible mrd + 3 years
elif mrd:
    # mrd is eligible now
else:
    # no recent convictions or dismissals
```

Comment (KentShikama) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: The code also needs a check to see whether `mrc.case != second_mrc.case`. It feels like this logic should be grouped together with `_check_mrc_time_eligibility`, otherwise there will be a lot of duplicated code.

Comment (wittejm) [reply src/backend/expungeservice/expunger/analyzers/time_analyzer.py]: I think refactoring this whole thing is probably a good idea, but let's definitely get this fix in so we can include it in a new version push this week. 

While at it there's probably a way to overhaul the way violations are handled with the mrc/second_mrc so we don't need edge case fixes for these new/old/old bugs: #718 #681 #654
